src/cpp/src/whisper_pipeline_static.cpp

```diff
@@ -0,0 +1,362 @@
+        }
+        std::string with_past_input_name =
+            std::regex_replace(source_output_name, std::regex("present"), "past");
+        dest.set_tensor(with_past_input_name, source.get_tensor(source_output_name));
+    }
+}
+
+void update_past_key_value(ov::InferRequest& source,
+                           ov::InferRequest& dest,
+                           const size_t kv_pos = 0u) {
+    // NB: Source outputs:
+    // present_key_values.0.decoder.key
+    // present_key_values.0.decoder.value
+
+    // NB: Dest inputs:
+    // past_key_values.0.decoder.key
+    // past_key_values.0.decoder.value
+
+    for (auto& source_output : source.get_compiled_model().outputs()) {
+        std::string source_output_name = source_output.get_any_name();
+        if (source_output_name.find("decoder") == std::string::npos) {
+            continue;
+        }
+
+        std::string with_past_input_name =
+            std::regex_replace(source_output_name, std::regex("present"), "past");
+
+        auto src_kv_tensor = source.get_tensor(source_output_name);
+        auto dst_kv_tensor = dest.get_tensor(with_past_input_name);
+        auto kv_size = src_kv_tensor.get_shape()[2];
+        // NB: Copy src_kv_tensor into dst_kv_tensor[:, :, kv_pos:kv_pos+kv_size, :]
+        auto dst_kv_tensor_slice = make_tensor_slice(dst_kv_tensor, 2u, kv_pos, kv_pos + kv_size);
+        src_kv_tensor.copy_to(dst_kv_tensor_slice);
+    }
+}
+
+int64_t decode(ov::Tensor& encoder_hidden_state,
+               ov::InferRequest& decoder,
+               const std::vector<int32_t>& input_ids,
+               const ov::genai::WhisperGenerationConfig& config,
+               bool do_suppress_tokens = true) {
+    // NB: Fill decoder inputs
+    encoder_hidden_state.copy_to(decoder.get_tensor("encoder_hidden_states"));
+    copy_to_tensor(input_ids, decoder.get_tensor("input_ids"));
+    fill_tensor<ov::float16>(decoder.get_tensor("attention_mask"), 1);
+
+    decoder.infer();
+
+    auto output_tensor = decoder.get_tensor("logits");
+    if (do_suppress_tokens) {
+        suppress_tokens(output_tensor, 0, config.begin_suppress_tokens);
+        suppress_tokens(output_tensor, 0, config.suppress_tokens);
+    }
+    int64_t output_token = ov::genai::utils::argmax(output_tensor, 0);
+    return output_token;
+}
+
+int64_t decode_with_past(ov::InferRequest& decoder_with_past,
+                         const int64_t input_id,
+                         const int64_t position_id,
+                         const ov::genai::WhisperGenerationConfig& config) {
+
+    // FIXME: Avoid this cast to i32. Why it's not i64 precision in model?
+    decoder_with_past.get_tensor("input_ids").data<int32_t>()[0] = static_cast<int32_t>(input_id);
+    // FIXME: Avoid this cast to i32. Why it's not i64 precision in model?
+    decoder_with_past.get_tensor("position_ids").data<int32_t>()[0] = static_cast<int32_t>(position_id);
+    // FIXME: Is "attention_mask" supposed to be f16?
```

Comment:
Mask is used later for computation. If mask is already in FP16, this avoids extra conversion step on device.